lib/services/local/headlessauthn_watcher.go

```diff
@@ -305,60 +317,67 @@ func (h *HeadlessAuthenticationWatcher) unassignSubscriber(i int) {
-	// name is the name of the headless authentication resource being subscribed to.
+	// name is a headless authentication name.
 	name string
+	// username is a teleport username.
+	username string
 	// updates is a channel used by the watcher to send resource updates. This channel
 	// will either be empty or have the latest update in its buffer.
-	updates   chan *types.HeadlessAuthentication
-	updatesMu sync.Mutex
+	updates chan *types.HeadlessAuthentication
 	// closed is a channel used to determine if the subscriber is closed.
-	closed chan struct{}
-}
-
-func (s *headlessAuthenticationSubscriber) Name() string {
-	return s.name
+	closed    chan struct{}
+	closeOnce sync.Once
 }
 
+// Updates is a channel used by the watcher to send headless authentication updates.
 func (s *headlessAuthenticationSubscriber) Updates() <-chan *types.HeadlessAuthentication {
 	return s.updates
 }
 
-func (s *headlessAuthenticationSubscriber) update(ha *types.HeadlessAuthentication) {
-	s.updatesMu.Lock()
-	defer s.updatesMu.Unlock()
-
-	// Drain stale update if there is one.
-	select {
-	case <-s.updates:
-	default:
-	}
-
-	s.updates <- apiutils.CloneProtoMsg(ha)
-}
-
-func (s *headlessAuthenticationSubscriber) Close() {
-	close(s.closed)
-}
-
-// WaitForUpdate waits until the headless authentication with the given name is updated in the
-// backend to meet the given condition or returns early if the condition results in an
-// error or if the watcher or given context is closed.
-func (h *HeadlessAuthenticationWatcher) WaitForUpdate(ctx context.Context, subscriber HeadlessAuthenticationSubscriber, cond func(*types.HeadlessAuthentication) (bool, error)) (*types.HeadlessAuthentication, error) {
+// WaitForUpdate returns the first update which passes the given condition, or returns
+// early if the condition results in an error or if the subscriber or given context is closed.
+func (s *headlessAuthenticationSubscriber) WaitForUpdate(ctx context.Context, cond func(*types.HeadlessAuthentication) (bool, error)) (*types.HeadlessAuthentication, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	for {
 		select {
-		case ha := <-subscriber.Updates():
+		case ha, ok := <-s.Updates():
+			if !ok {
+				return nil, ErrHeadlessAuthenticationWatcherClosed
+			}
 			if ok, err := cond(ha); err != nil {
 				return nil, trace.Wrap(err)
 			} else if ok {
 				return ha, nil
 			}
 		case <-ctx.Done():
 			return nil, trace.Wrap(ctx.Err())
-		case <-h.Done():
+		case <-s.Done():
 			return nil, ErrHeadlessAuthenticationWatcherClosed
 		}
 	}
 }
+
+// Done returns a channel that's closed when the subscriber is closed.
+func (s *headlessAuthenticationSubscriber) Done() <-chan struct{} {
+	return s.closed
+}
+
+// Close closes the subscriber and its channels. This frees up resources for the watcher
+// and should always be called on completion.
+func (s *headlessAuthenticationSubscriber) Close() {
+	s.closeOnce.Do(func() {
+		close(s.closed)
+	})
+}
```

Comment:
nit: maybe add `close(s.updates)` here and remove the `close(subscriber.updates)` after sub is unassigned by the watcher?

Reply:
The issue is that the watcher acts separately from the subscriber. We don't want to close the updates channel at the same time as the watcher tries to send to the updates channel. This was my solution to it, maybe there is a better way though, wdyt?

Reply:
I made some changes to this logic: 214b41c4957095c71805348cd537204b8650d7c6

Reply:
sorry I missed this feedback when I looked earlier.

I see what you mean now wrt to the original code. I think the new code will deadlock though:
1. call `watcher.close()`
2. `watcher.close()` loops through subscribers calling `sub.Close()`
3. `sub.onClose` callback calls `watcher.unassign`
4. `watcher.unassign` tries to acquire the lock

I leave it up to you if you want to just revert to the previous code, which worked fine. I would probably only change it so that the subscriber "owns" its update channel, and closes it within `sub.Close()` - then changing the original goroutine you had like this:

```go
// Subscribe creates a subscriber for a specific headless authentication.
func (h *HeadlessAuthenticationWatcher) Subscribe(ctx context.Context, username, name string) (HeadlessAuthenticationSubscriber, error) {
	if name == "" {
		return nil, trace.BadParameter("name must be provided")
	}
	if username == "" {
		return nil, trace.BadParameter("username must be provided")
	}

	i, err := h.assignSubscriber(username, name)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	subscriber := h.subscribers[i]

	go func() {
		select {
		case <-subscriber.Done():
			h.unassignSubscriber(i)
		case <-h.Done():
			h.unassignSubscriber(i)
			// reclaim the subscriber and close its updates channel.
			subscriber.Close()
		}
	}()
```